### DISCUSSION TITLE
[Next 13.4] Middleware acts like a client component

### DISCUSSION THREAD

**Discussion by SplittyDev** [vercel/next.js#49246] [Help] [answered]
Hi, I just upgraded to Next.JS 13.4 (App Router) and created an `src/middleware.ts` file.

Since I'm using the middleware for validating the user's session, I'm importing `cookies` from `next/headers`. When running, NextJS prints an error (line breaks inserted for legibility):

```
You're importing a component that needs next/headers.

That only works in a Server Component but one of its parents is marked
with "use client", so it's a Client Component.

   ,-[/Users/redacted/Developer/redacted/src/middleware.ts:1:1]
 1 | import { cookies } from 'next/headers'
   : ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 2 | import { NextResponse, type NextRequest } from 'next/server'
   `----

One of these is marked as a client entry with "use client":
  ./src/middleware.ts
```

Based on the docs[^1], shouldn't middleware run on the server, since it can send a different `NextResponse`?
I'm not entirely sure that this is a bug, but it's certainly surprising behavior.

---

Digging a bit deeper, this only seems to happen when sending any kind of HTTP request.

At first, I thought that `axios` might be the issue, but replacing it with a `fetch` call yields the same result.
Removing the request completely makes the middleware work as expected.

Since server components usually allow requests, this is even more surprising.

---

<details>
<summary>If anyone is interested in the code, here's a minimal example</summary>

```typescript
import { cookies } from 'next/headers'
import { NextResponse, type NextRequest } from 'next/server'

export async function middleware(request: NextRequest) {

  // Redirect to login if session cookie is not present
  if (!cookies().has('session')) {
    return NextResponse.redirect(new URL('/login', request.url))
  }

  // Check session validity with the server
  const session = cookies().get('session')
  const res = await fetch(`https://example.com/session`, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${session}`
    },
    cache: 'no-store',
  })

  // Redirect to login if session is invalid
  if (res.status === 401) {
    return NextResponse.redirect(new URL('/login', request.url))
  }

  return NextResponse.next()
}

export const config = {
  matcher: '/:path*'
}
```

</details>

[^1]: https://nextjs.org/docs/app/building-your-application/routing/middleware

**Comment by SplittyDev** [comment, marked as answer]
Turns out it wasn't because of the fetch call. It just doesn't seem to be possible to import `next/headers` in middleware at all, but there's a `cookies` object on `NextRequest`.

**Reply by Utkarshvr** [reply]
**But why can't we import `next/headers` in middleware?**
*Is there any fix?*

**Reply by ceduth** [reply]
Hi, check this please:

[Modifying Request Headers in Middleware](https://vercel.com/templates/next.js/edge-functions-modify-request-header).
